### PR TITLE
Fix HTTP/3 bug: uploading payload larger than 65536 will be blocked until timeout.

### PR DESCRIPTION
**Pull request by MegatronKing**
The upload buffer default size is 65536, if we post a payload larger than this, the first buffer is ok, but when sending the second buffer, the stream will be blocked with a curlcode CURLE_AGAIN. I supoose the cause is the upload_len was not reset.

### CODE REVIEW

File: lib/vquic/ngtcp2.c [hunk 1]
```diff
@@ -1660,6 +1660,12 @@ static ssize_t ngh3_stream_send(struct Curl_easy *data,
     return -1;
   }
 
+  // Reset post upload buffer after resumed.
```
Comment (bagder) [inline lib/vquic/ngtcp2.c:1663]: This style of comment is not allowed in curl code as its not C89 compliant! Use /* */ ...

File: lib/vquic/ngtcp2.c [hunk 2]
```diff
@@ -1660,6 +1660,12 @@ static ssize_t ngh3_stream_send(struct Curl_easy *data,
...
+  if(data->set.postfields) {
```
Comment (bagder) [inline lib/vquic/ngtcp2.c:1664]: This seems too specific for POST. Shouldn't this rather check if `stream->upload_mem` is set?

**Review by bagder — CHANGES_REQUESTED** [review]

**Comment by bagder** [comment]
and the other comment, what do you think about it?

**Comment by MegatronKing** [comment]
@bagder 
Actually, I am not very confident about my fix, I tested with all my POST cases and all approved.
I am confusing about the function 'cb_h3_readfunction', when and how it would be invoked? And this line https://github.com/curl/curl/blob/b836b065bd6598a8c549863825fcfcd0306b3cb5/lib/vquic/ngtcp2.c#L1350 also is specific for POST and I am not able to understand too.

**Comment by MegatronKing** [comment]
My test cases:
- POST application/octem 10/100/1024/1024 * 1024/10 * 1024 * 1024
- POST multipart/form-data 10/100/1024/1024 * 1024/10 * 1024 * 1024
And something intersting, POST  multipart/form-data and _data->set.postfields_ is false.

**Comment by bagder** [comment]
> POST multipart/form-data and data->set.postfields is false

That's why I didn't like the fix. `set.postfields` is only set with `CURLOPT_POSTFIELDS`  while you can send data over HTTP several other ways with libcurl. One of those ways you see. You can also try `curl -T [file] -X POST $URL` ...

So if you make the change I proposed, does it still work or isn't that good enough?

**Comment by MegatronKing** [comment]
if data->set.postfields is true, a flag NGHTTP3_DATA_FLAG_EOF is set and return, and the value of stream->upload_len would never be updated, right? I think this is a bug.

**Comment by MegatronKing** [comment]
Agree with the fix is not good enough, I updated!

**Review by bagder — APPROVED** [review]

**Comment by bagder** [comment]
do you have a HTTP/3 test server somewhere that you can do large posts to, to verify this fix?

**Comment by bagder** [comment]
Thanks, I pushed this as I believe it does improve things. I still experience problems if I upload multiple megabytes, but after all the data has been sent...

**Pull request closed without merging**